managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java

```diff
@@ -1504,6 +1504,8 @@ public synchronized void updateLedgersIdsComplete(Stat stat) {
                 if (existsOp.ledger != null) {
                     existsOp.close();
                     existsOp = OpAddEntry.create(existsOp.ml, existsOp.data, existsOp.getNumberOfMessages(), existsOp.callback, existsOp.ctx);
+                    // release the extra retain
+                    ReferenceCountUtil.release(existsOp.data);
```

Comment:
Just one question, doesn't `existsOp.close` release the reference? Or can you clarify (potentially comment in the code) why do we need to release extra retain here?

Reply:
> Just one question, doesn't existsOp.close release the reference?

No it doesn't.

> why do we need to release extra retain here?

OpAddEntry.create will call .retain , that's the extra retain. I already tried to clarify in the comment in the code. Does it need more clarification?

Reply:
LGTM, just add more context here. We can safe to release the buffer here because we closed the old OpAddEntry, if the OpAddEntry in the closed state, even if the `addComplete` been called, will not use the buffer again.

```
        if (!STATE_UPDATER.compareAndSet(OpAddEntry.this, State.INITIATED, State.COMPLETED)) {
            log.warn("[{}] The add op is terminal legacy callback for entry {}-{} adding.", ml.getName(), lh.getId(), entryId);
            OpAddEntry.this.recycle();
            return;
        }
```